pkg/chunk/cached_store.go

```diff
@@ -755,6 +770,94 @@ func parseObjOrigSize(key string) int {
 	return l
 }
 
+func (store *cachedStore) uploadStagingFile(key string, stagingPath string) int {
+	block, err := ioutil.ReadFile(stagingPath)
+	if err != nil {
+		logger.Errorf("open %s: %s", stagingPath, err)
+		return 0
+	}
+	buf := make([]byte, store.compressor.CompressBound(len(block)))
+	n, err := store.compressor.Compress(buf, block)
+	if err != nil {
+		logger.Errorf("compress chunk %s: %s", stagingPath, err)
+		return 0
+	}
+	compressed := buf[:n]
+
+	if strings.Count(key, "_") == 1 {
+		// add size at the end
+		key = fmt.Sprintf("%s_%d", key, len(block))
+	}
+	try := 0
+	for {
+		err := store.storage.Put(key, bytes.NewReader(compressed))
+		if err == nil {
+			break
+		}
+		logger.Infof("upload %s: %s (try %d)", key, err, try)
+		try++
+		time.Sleep(time.Second * time.Duration(try*try))
+	}
+	store.bcache.uploaded(key, len(block))
+	os.Remove(stagingPath)
+	return len(block)
+}
+
+func computeFreeGoal(innerStore *cacheStore, br float32) int64 {
+	total, _, _, _ := getDiskUsage(innerStore.dir)
+	goal := innerStore.capacity * 80 / 100
+	toFree := int64(float32(total) * (innerStore.freeRatio - br))
+	if toFree > innerStore.used {
+		goal = 0
+	} else if innerStore.used-toFree < goal {
+		goal = innerStore.used - toFree
+	}
+	return goal
+}
+
+func (store *cachedStore) cleanDelayStaging() {
+	cacheStoreKeys := make(map[*cacheStore][]string)
+
+	for key, _ := range store.pendingKeys {
+		innerStore := store.bcache.getStore(key)
+		cacheStoreKeys[innerStore] = append(cacheStoreKeys[innerStore], key)
+	}
+
+	for innerStore, keys := range cacheStoreKeys {
+		go func() {
+			br, fr := innerStore.curFreeRatio()
+			if br >= innerStore.freeRatio && fr >= innerStore.freeRatio {
+				return
+			}
+			goal := computeFreeGoal(innerStore, br)
+			var freed int
+			for _, key := range keys {
+				store.currentUpload <- true
+				go func(key string, stagingPath string) {
+					defer func() {
+						<-store.currentUpload
+					}()
+
+					defer func() {
+						store.pendingMutex.Lock()
+						delete(store.pendingKeys, key)
+						store.pendingMutex.Unlock()
+					}()
+					fileInfo := innerStore.keys[key]
+					isTimeout := time.Now().Unix()-int64(fileInfo.atime) > int64(store.conf.DelayUpload.Seconds())
+					if isTimeout {
+						freed += store.uploadStagingFile(key, stagingPath)
+					}
+
+				}(key, innerStore.stagePath(key))
+				if int64(freed) >= goal {
```

Comment:
it looks like **goal** is the target space used, which is not comparable with **freed**